File: RAGTest/src/api/folders.py
"""
폴더 관리 API 라우터
"""
from flask import Blueprint, request, jsonify
from src.folder_manager import FolderManager

# Blueprint 생성
folders_bp = Blueprint('folders', __name__, url_prefix='/api/folders')

# FolderManager 인스턴스 (app.py에서 주입받을 예정)
folder_manager = None


def init_folder_manager(manager: FolderManager):
    """FolderManager 인스턴스 주입"""
    global folder_manager
    folder_manager = manager


@folders_bp.route('', methods=['GET'])
def get_folders():
    """전체 폴더 트리 조회"""
    try:
        tree = folder_manager.get_folder_tree()
        return jsonify({
            'success': True,
            'folders': tree
        })
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@folders_bp.route('/list', methods=['GET'])
def get_all_folders():
    """모든 폴더 평면 목록 조회"""
    try:
        folders = folder_manager.get_all_folders()
        return jsonify({
            'success': True,
            'folders': folders
        })
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@folders_bp.route('', methods=['POST'])
def create_folder():
    """새 폴더 생성"""
    try:
        data = request.get_json()
        name = data.get('name')
        parent_id = data.get('parent_id')
        color = data.get('color', '#6366f1')
        icon = data.get('icon', 'folder')

        if not name:
            return jsonify({
                'success': False,
                'error': '폴더 이름이 필요합니다.'
            }), 400

        folder = folder_manager.create_folder(name, parent_id, color, icon)

        return jsonify({
            'success': True,
            'folder': folder
        })
    except ValueError as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 400
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@folders_bp.route('/<folder_id>', methods=['GET'])
def get_folder(folder_id):
    """특정 폴더 조회"""
    try:
        folder = folder_manager.get_folder(folder_id)

        if folder:
            return jsonify({
                'success': True,
                'folder': folder
            })
        else:
            return jsonify({
                'success': False,
                'error': '폴더를 찾을 수 없습니다.'
            }), 404
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@folders_bp.route('/<folder_id>', methods=['PUT'])
def update_folder(folder_id):
    """폴더 수정"""
    try:
        data = request.get_json()
        name = data.get('name')
        color = data.get('color')
        icon = data.get('icon')

        folder = folder_manager.update_folder(folder_id, name, color, icon)

        return jsonify({
            'success': True,
            'folder': folder
        })
    except ValueError as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 404
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@folders_bp.route('/<folder_id>', methods=['DELETE'])
def delete_folder(folder_id):
    """폴더 삭제"""
    try:
        recursive = request.args.get('recursive', 'false').lower() == 'true'
        success = folder_manager.delete_folder(folder_id, recursive=recursive)

        if success:
            return jsonify({
                'success': True,
                'message': '폴더가 삭제되었습니다.'
            })
        else:
            return jsonify({
                'success': False,
                'error': '폴더를 찾을 수 없습니다.'
            }), 404
    except ValueError as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 400
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@folders_bp.route('/<folder_id>/move', methods=['POST'])
def move_folder(folder_id):
    """폴더 이동"""
    try:
        data = request.get_json()
        new_parent_id = data.get('new_parent_id')  # None이면 루트로

        folder = folder_manager.move_folder(folder_id, new_parent_id)

        return jsonify({
            'success': True,
            'folder': folder
        })
    except ValueError as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 400
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@folders_bp.route('/<folder_id>/documents', methods=['GET'])
def get_folder_documents(folder_id):
    """폴더 내 문서 목록 조회"""
    try:
        include_subfolders = request.args.get('include_subfolders', 'false').lower() == 'true'
        documents = folder_manager.get_documents_in_folder(folder_id, include_subfolders)

        return jsonify({
            'success': True,
            'folder_id': folder_id,
            'documents': documents,
            'count': len(documents)
        })
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@folders_bp.route('/<folder_id>/documents', methods=['POST'])
def assign_document_to_folder(folder_id):
    """문서를 폴더에 할당"""
    try:
        data = request.get_json()
        doc_id = data.get('doc_id')

        if not doc_id:
            return jsonify({
                'success': False,
                'error': '문서 ID가 필요합니다.'
            }), 400

        success = folder_manager.assign_document_to_folder(doc_id, folder_id)

        if success:
            return jsonify({
                'success': True,
                'message': f'문서가 폴더에 추가되었습니다.'
            })
        else:
            return jsonify({
                'success': False,
                'error': '폴더를 찾을 수 없습니다.'
            }), 404
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@folders_bp.route('/<folder_id>/documents/<path:doc_id>', methods=['DELETE'])
def remove_document_from_folder(folder_id, doc_id):
    """문서를 폴더에서 제거"""
    try:
        success = folder_manager.remove_document_from_folder(doc_id)

        if success:
            return jsonify({
                'success': True,
                'message': '문서가 폴더에서 제거되었습니다.'
            })
        else:
            return jsonify({
                'success': False,
                'error': '문서를 찾을 수 없습니다.'
            }), 404
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@folders_bp.route('/<folder_id>/path', methods=['GET'])
def get_folder_path(folder_id):
    """폴더 경로 조회 (브레드크럼용)"""
    try:
        path = folder_manager.get_folder_path(folder_id)

        return jsonify({
            'success': True,
            'path': path
        })
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@folders_bp.route('/stats', methods=['GET'])
def get_folder_stats():
    """폴더 통계"""
    try:
        stats = folder_manager.get_folder_stats()
        return jsonify({
            'success': True,
            'stats': stats
        })
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@folders_bp.route('/document/<path:doc_id>', methods=['GET'])
def get_document_folder(doc_id):
    """문서가 속한 폴더 조회"""
    try:
        folder_id = folder_manager.get_document_folder(doc_id)

        if folder_id:
            folder = folder_manager.get_folder(folder_id)
            return jsonify({
                'success': True,
                'folder_id': folder_id,
                'folder': folder
            })
        else:
            return jsonify({
                'success': True,
                'folder_id': None,
                'folder': None
            })
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

File: rag-microservices/services/web-ui/static/js/modules/features/manage/folders.js
// ===================================================================
// 폴더 관리 모듈
// 문서를 폴더로 정리하는 기능 제공
// ===================================================================

/**
 * 현재 선택된 폴더 ID
 * @type {string|null}
 */
let selectedFolderId = null;

/**
 * 폴더 데이터 캐시
 * @type {Array}
 */
let foldersCache = [];

/**
 * 문서 필터 콜백
 * @type {Function|null}
 */
let onFolderFilterCallback = null;

/**
 * 폴더 필터 콜백 설정
 * @param {Function} callback - 폴더 선택 시 호출될 콜백
 */
export function setFolderFilterCallback(callback) {
    onFolderFilterCallback = callback;
}

/**
 * 폴더 트리 로드 및 렌더링
 */
export async function loadFolders() {
    try {
        const response = await fetch('api/folders');
        const data = await response.json();

        if (data.success) {
            foldersCache = data.folders;
            renderFolderTree(data.folders);
        }
    } catch (error) {
        console.error('폴더 로드 실패:', error);
    }
}

/**
 * 폴더 트리 렌더링
 * @param {Array} folders - 폴더 트리 데이터
 */
function renderFolderTree(folders) {
    const container = document.getElementById('folder-tree');
    if (!container) return;

    container.innerHTML = '';

    // "전체 문서" 항목
    const allItem = createFolderItem({
        id: null,
        name: '전체 문서',
        icon: 'files',
        document_count: getTotalDocumentCount(folders)
    }, true);
    container.appendChild(allItem);

    // 폴더 트리 렌더링
    folders.forEach(folder => {
        const folderElement = createFolderElement(folder, 0);
        container.appendChild(folderElement);
    });

    // "미분류" 항목
    const unassignedItem = createFolderItem({
        id: '_unassigned',
        name: '미분류',
        icon: 'file-question',
        document_count: 0
    }, false);
    container.appendChild(unassignedItem);

    // 새 폴더 버튼
    const addBtn = document.createElement('button');
    addBtn.className = 'folder-add-btn';
    addBtn.innerHTML = '<span>+</span> 새 폴더';
    addBtn.onclick = () => showCreateFolderDialog();
    container.appendChild(addBtn);
}

/**
 * 전체 문서 수 계산
 * @param {Array} folders - 폴더 목록
 * @returns {number}
 */
function getTotalDocumentCount(folders) {
    let total = 0;
    folders.forEach(folder => {
        total += folder.document_count || 0;
        if (folder.children_data) {
            total += getTotalDocumentCount(folder.children_data);
        }
    });
    return total;
}

/**
 * 폴더 요소 생성 (재귀)
 * @param {Object} folder - 폴더 데이터
 * @param {number} depth - 깊이
 * @returns {HTMLElement}
 */
function createFolderElement(folder, depth) {
    const item = createFolderItem(folder, false, depth);

    // 하위 폴더가 있는 경우
    if (folder.children_data && folder.children_data.length > 0) {
        const children = document.createElement('div');
        children.className = 'folder-children';
        children.style.display = 'none';

        folder.children_data.forEach(child => {
            children.appendChild(createFolderElement(child, depth + 1));
        });

        item.appendChild(children);
    }

    return item;
}

/**
 * 폴더 항목 생성
 * @param {Object} folder - 폴더 데이터
 * @param {boolean} isAllDocs - "전체 문서" 항목인지
 * @param {number} depth - 깊이
 * @returns {HTMLElement}
 */
function createFolderItem(folder, isAllDocs = false, depth = 0) {
    const item = document.createElement('div');
    item.className = 'folder-item';
    if (selectedFolderId === folder.id) {
        item.classList.add('selected');
    }
    item.style.paddingLeft = `${12 + depth * 16}px`;
    item.dataset.folderId = folder.id;

    const hasChildren = folder.children_data && folder.children_data.length > 0;

    // 폴더 아이콘
    const icon = getIconForFolder(folder, hasChildren);

    item.innerHTML = `
        <span class="folder-toggle ${hasChildren ? '' : 'hidden'}" onclick="event.stopPropagation(); toggleFolder('${folder.id}')">
            ${hasChildren ? '▶' : ''}
        </span>
        <span class="folder-icon" style="color: ${folder.color || '#6366f1'}">${icon}</span>
        <span class="folder-name">${folder.name}</span>
        <span class="folder-count">${folder.document_count || 0}</span>
    `;

    // 클릭 이벤트
    item.onclick = (e) => {
        if (!e.target.classList.contains('folder-toggle')) {
            selectFolder(folder.id);
        }
    };

    // 우클릭 컨텍스트 메뉴 (시스템 폴더 제외)
    if (!isAllDocs && folder.id !== '_unassigned' && !folder.is_system) {
        item.oncontextmenu = (e) => {
            e.preventDefault();
            showFolderContextMenu(e, folder);
        };
    }

    return item;
}

/**
 * 폴더 아이콘 반환
 * @param {Object} folder - 폴더 데이터
 * @param {boolean} hasChildren - 하위 폴더 여부
 * @returns {string}
 */
function getIconForFolder(folder, hasChildren) {
    if (folder.icon === 'files') return '📑';
    if (folder.icon === 'file-question') return '📄';
    if (folder.icon === 'cloud') return '☁️';
    if (folder.icon === 'building') return '🏢';
    if (folder.icon === 'landmark') return '🏛️';
    return hasChildren ? '📂' : '📁';
}

/**
 * 폴더 펼침/접기
 * @param {string} folderId - 폴더 ID
 */
export function toggleFolder(folderId) {
    const item = document.querySelector(`.folder-item[data-folder-id="${folderId}"]`);
    if (!item) return;

    const children = item.querySelector('.folder-children');
    const toggle = item.querySelector('.folder-toggle');

    if (children) {
        const isOpen = children.style.display !== 'none';
        children.style.display = isOpen ? 'none' : 'block';
        toggle.textContent = isOpen ? '▶' : '▼';
    }
}
window.toggleFolder = toggleFolder;

/**
 * 폴더 선택
 * @param {string|null} folderId - 폴더 ID (null이면 전체)
 */
export function selectFolder(folderId) {
    // 이전 선택 해제
    document.querySelectorAll('.folder-item.selected').forEach(el => {
        el.classList.remove('selected');
    });

    // 새 선택
    selectedFolderId = folderId;
    const item = document.querySelector(`.folder-item[data-folder-id="${folderId}"]`);
    if (item) {
        item.classList.add('selected');
    }

    // 콜백 호출
    if (onFolderFilterCallback) {
        onFolderFilterCallback(folderId);
    }
}

/**
 * 현재 선택된 폴더 ID 반환
 * @returns {string|null}
 */
export function getSelectedFolderId() {
    return selectedFolderId;
}

/**
 * 새 폴더 생성 다이얼로그
 * @param {string|null} parentId - 부모 폴더 ID
 */
export function showCreateFolderDialog(parentId = null) {
    const name = prompt('새 폴더 이름을 입력하세요:');
    if (!name || !name.trim()) return;

    createFolder(name.trim(), parentId);
}

/**
 * 폴더 생성 API 호출
 * @param {string} name - 폴더 이름
 * @param {string|null} parentId - 부모 폴더 ID
 */
async function createFolder(name, parentId = null) {
    try {
        const response = await fetch('api/folders', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                name: name,
                parent_id: parentId
            })
        });

        const data = await response.json();
        if (data.success) {
            loadFolders();
        } else {
            alert('폴더 생성 실패: ' + data.error);
        }
    } catch (error) {
        console.error('폴더 생성 실패:', error);
        alert('폴더 생성 중 오류가 발생했습니다.');
    }
}

/**
 * 폴더 컨텍스트 메뉴 표시
 * @param {MouseEvent} e - 마우스 이벤트
 * @param {Object} folder - 폴더 데이터
 */
function showFolderContextMenu(e, folder) {
    // 기존 메뉴 제거
    const existingMenu = document.querySelector('.folder-context-menu');
    if (existingMenu) existingMenu.remove();

    const menu = document.createElement('div');
    menu.className = 'folder-context-menu';
    menu.style.left = `${e.pageX}px`;
    menu.style.top = `${e.pageY}px`;

    menu.innerHTML = `
        <div class="context-menu-item" onclick="renameFolderPrompt('${folder.id}', '${folder.name}')">
            <span>📝</span> 이름 바꾸기
        </div>
        <div class="context-menu-item" onclick="showCreateFolderDialogWithParent('${folder.id}')">
            <span>📁</span> 하위 폴더 추가
        </div>
        <div class="context-menu-separator"></div>
        <div class="context-menu-item danger" onclick="deleteFolderPrompt('${folder.id}', '${folder.name}')">
            <span>🗑️</span> 삭제
        </div>
    `;

    document.body.appendChild(menu);

    // 클릭 시 메뉴 닫기
    const closeMenu = () => {
        menu.remove();
        document.removeEventListener('click', closeMenu);
    };
    setTimeout(() => document.addEventListener('click', closeMenu), 0);
}

/**
 * 폴더 이름 변경 프롬프트
 * @param {string} folderId - 폴더 ID
 * @param {string} currentName - 현재 이름
 */
window.renameFolderPrompt = async function(folderId, currentName) {
    const newName = prompt('새 이름을 입력하세요:', currentName);
    if (!newName || newName.trim() === currentName) return;

    try {
        const response = await fetch(`api/folders/${folderId}`, {
            method: 'PUT',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ name: newName.trim() })
        });

        const data = await response.json();
        if (data.success) {
            loadFolders();
        } else {
            alert('이름 변경 실패: ' + data.error);
        }
    } catch (error) {
        console.error('이름 변경 실패:', error);
    }
};

/**
 * 하위 폴더 생성 다이얼로그
 * @param {string} parentId - 부모 폴더 ID
 */
window.showCreateFolderDialogWithParent = function(parentId) {
    showCreateFolderDialog(parentId);
};

/**
 * 폴더 삭제 확인
 * @param {string} folderId - 폴더 ID
 * @param {string} folderName - 폴더 이름
 */
window.deleteFolderPrompt = async function(folderId, folderName) {
    if (!confirm(`'${folderName}' 폴더를 삭제하시겠습니까?`)) return;

    try {
        const response = await fetch(`api/folders/${folderId}?recursive=true`, {
            method: 'DELETE'
        });

        const data = await response.json();
        if (data.success) {
            if (selectedFolderId === folderId) {
                selectFolder(null);
            }
            loadFolders();
        } else {
            alert('삭제 실패: ' + data.error);
        }
    } catch (error) {
        console.error('삭제 실패:', error);
    }
};

/**
 * 문서를 폴더에 할당
 * @param {string} docId - 문서 ID
 * @param {string} folderId - 폴더 ID
 * @param {string} collection - 컬렉션명 (기본값: documents)
 */
export async function assignDocumentToFolder(docId, folderId, collection = 'documents') {
    try {
        const response = await fetch(`api/folders/${folderId}/documents`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ doc_id: docId, collection: collection })
        });

        const data = await response.json();
        if (data.success) {
            loadFolders();
            return true;
        }
        return false;
    } catch (error) {
        console.error('폴더 할당 실패:', error);
        return false;
    }
}

/**
 * 문서를 폴더에서 제거
 * @param {string} docId - 문서 ID
 * @param {string} folderId - 폴더 ID
 */
export async function removeDocumentFromFolder(docId, folderId) {
    try {
        const response = await fetch(`api/folders/${folderId}/documents/${encodeURIComponent(docId)}`, {
            method: 'DELETE'
        });

        const data = await response.json();
        if (data.success) {
            loadFolders();
            return true;
        }
        return false;
    } catch (error) {
        console.error('폴더에서 제거 실패:', error);
        return false;
    }
}

/**
 * 문서의 폴더 정보 조회
 * @param {string} docId - 문서 ID
 * @returns {Promise<Object|null>}
 */
export async function getDocumentFolder(docId) {
    try {
        const response = await fetch(`api/folders/document/${encodeURIComponent(docId)}`);
        const data = await response.json();

        if (data.success) {
            return data.folder;
        }
        return null;
    } catch (error) {
        console.error('문서 폴더 조회 실패:', error);
        return null;
    }
}

/**
 * 폴더 내 문서 ID 목록 조회
 * @param {string} folderId - 폴더 ID
 * @param {boolean} includeSubfolders - 하위 폴더 포함 여부
 * @returns {Promise<string[]>}
 */
export async function getDocumentsInFolder(folderId, includeSubfolders = false) {
    try {
        const url = `api/folders/${folderId}/documents?include_subfolders=${includeSubfolders}`;
        const response = await fetch(url);
        const data = await response.json();

        if (data.success) {
            return data.documents;
        }
        return [];
    } catch (error) {
        console.error('폴더 문서 조회 실패:', error);
        return [];
    }
}
